Lib/test/support/os_helper.py

```diff
@@ -294,6 +294,33 @@ def skip_unless_working_chmod(test):
     return test if ok else unittest.skip(msg)(test)
 
 
+@contextlib.contextmanager
+def save_mode(path, *, quiet=False):
+    """Context manager that restores the mode (permissions) of path on exit.
+
+    Arguments:
+
+      path: Path of the file to restore mode of.
+
+      quiet: if False (the default), the context manager raises an exception
+        on error.  Otherwise, it issues only a warning and keeps the current
+        working directory the same.
+
+    """
+    saved_mode = os.stat(path)
+    try:
+        yield
+    finally:
+        try:
+            os.chmod(path, saved_mode.st_mode)
+        except OSError as exc:
+            if not quiet:
+                raise
+            warnings.warn(f'tests may fail, unable to restore the mode of '
+                          f'{path!r} to {saved_mode.st_mode}: {exc}',
+                          RuntimeWarning, stacklevel=3)
+
+
```

Comment:
I appreciate you thinking about `save_mode` but that leaves the caller to do half of the work (initial chmod, including error handling) but not the cleanup. Perhaps better would be a `change_mode` that (a) invokes `skip_unless_working_chmod` and (b) takes as a parameter the mode to set. This approach should also simplify the usage in the test.

```suggestion
def change_mode(test, path, mode, *, quiet=False):
    """Context manager that restores the mode (permissions) of path on exit.

    Arguments:

      test: ...
      
      path: Path of the file to restore mode of.
      
      mode: ...

      quiet: if False (the default), the context manager raises an exception
        on error.  Otherwise, it issues only a warning and keeps the current
        working directory the same.

    """
    skip_unless_working_chmod(test)
    
    saved_mode = os.stat(path)

    try:
        os.chmod(path, mode)
    except OSError as exc:
        if not quiet:
            raise
        warnings.warn(f'tests may fail, unable to change the mode '
                      f'of {path!r} to {mode!r}: {exc}',
                      RuntimeWarning, stacklevel=3)
    try:
        yield
    finally:
        os.chmod(path, saved_mode.st_mode)


```

This approach is more symmetric with `change_cwd` and saves the caller from having to deal with the consequences of the error. It may not even be valuable or necessary to handle the exception or solicit `quiet`, because the test should already have been skipped at that point.

```suggestion
def change_mode(test, path, mode):
    """Context manager that restores the mode (permissions) of path on exit.

    Arguments:

      test: ...
      
      path: Path of the file to restore mode of.
      
      mode: ...

    """
    skip_unless_working_chmod(test)
    
    saved_mode = os.stat(path)

    os.chmod(path, mode)

    try:
        yield
    finally:
        os.chmod(path, saved_mode.st_mode)


```